File: frontend/demo/component/select/select-custom-renderer-label.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/select';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import { getPeople } from 'Frontend/demo/domain/DataService';
import { applyTheme } from 'Frontend/demo/theme';
import type Person from 'Frontend/generated/com/vaadin/demo/domain/Person';

const formatPersonFullName = (person: Person) => `${person.firstName} ${person.lastName}`;

@customElement('select-custom-renderer-label')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  @state()
  private people: Person[] = [];

  protected override async firstUpdated() {
    this.people = (await getPeople({ count: 5 })).people;
  }

  protected override render() {
    return html`
      <vaadin-select label="Assignee">
        <vaadin-select-list-box slot="overlay">
          ${this.people.map(
            (person) => html`
              <!-- tag::snippet[] -->
              <!-- Use the label attribute to display full name of the person as selected value label -->
              <vaadin-select-item value="${person.id}" label="${formatPersonFullName(person)}">
                <div class="person-item">
                  <img
                    src="${person.pictureUrl}"
                    alt="Portrait of ${formatPersonFullName(person)}"
                    style="width: 2.25rem;"
                  />
                  <span>${formatPersonFullName(person)}</span>
                  <span>${person.profession}</span>
                </div>
              </vaadin-select-item>
              <!-- end::snippet[] -->
            `
          )}
        </vaadin-select-list-box>
      </vaadin-select>
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/select/SelectCustomRendererLabel.java
package com.vaadin.demo.component.select;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;
import com.vaadin.demo.domain.Person;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.Image;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.select.Select;
import com.vaadin.flow.data.renderer.ComponentRenderer;
import com.vaadin.flow.router.Route;

import java.util.List;

@Route("select-custom-renderer-label")
public class SelectCustomRendererLabel extends Div {

    public SelectCustomRendererLabel() {
        // tag::snippet[]
        Select<Person> select = new Select<>();
        select.setLabel("Assignee");
        // Use a custom renderer for items in the dropdown
        select.setRenderer(SelectCustomRendererLabel.createPersonRenderer());
        // Display full name of the person as selected value label
        select.setItemLabelGenerator(Person::getFullName);

        List<Person> people = DataService.getPeople(5);
        select.setItems(people);
        // end::snippet[]

        add(select);
    }

    private static ComponentRenderer<Div, Person> createPersonRenderer() {
        return new ComponentRenderer<>(person -> {
            Image image = new Image();
            image.setSrc(person.getPictureUrl());
            image.setAlt("Portrait of " + person.getFirstName() + " "
                    + person.getLastName());
            image.setWidth("2.25rem");

            Span name = new Span(person.getFullName());
            Span profession = new Span(person.getProfession());

            Div personItem = new Div(image, name, profession);
            personItem.addClassName("person-item");
            return personItem;
        });
    }

    public static class Exporter extends // hidden-source-line
            DemoExporter<SelectCustomRendererLabel> { // hidden-source-line
    } // hidden-source-line
}
